cluster-autoscaler/cloudprovider/digitalocean/digitalocean_cloud_provider.go

```diff
@@ -0,0 +1,193 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package digitalocean
+
+import (
+	"io"
+	"os"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+	"k8s.io/autoscaler/cluster-autoscaler/config"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+	"k8s.io/klog"
+)
+
+var _ cloudprovider.CloudProvider = (*digitaloceanCloudProvider)(nil)
+
+const (
+	// GPULabel is the label added to nodes with GPU resource.
+	GPULabel = "cloud.digitalocean.com/gpu-node"
+)
+
+// digitaloceanCloudProvider implements CloudProvider interface.
+type digitaloceanCloudProvider struct {
+	manager         *Manager
+	resourceLimiter *cloudprovider.ResourceLimiter
+}
+
+func newDigitalOceanCloudProvider(manager *Manager, rl *cloudprovider.ResourceLimiter) (*digitaloceanCloudProvider, error) {
+	if err := manager.Refresh(); err != nil {
+		return nil, err
+	}
+
+	return &digitaloceanCloudProvider{
+		manager:         manager,
+		resourceLimiter: rl,
+	}, nil
+}
+
+// Name returns name of the cloud provider.
+func (d *digitaloceanCloudProvider) Name() string {
+	return cloudprovider.DigitalOceanProviderName
+}
+
+// NodeGroups returns all node groups configured for this cloud provider.
+func (d *digitaloceanCloudProvider) NodeGroups() []cloudprovider.NodeGroup {
+	nodeGroups := make([]cloudprovider.NodeGroup, len(d.manager.nodeGroups))
+	for i, ng := range d.manager.nodeGroups {
+		nodeGroups[i] = ng
+	}
+	return nodeGroups
+}
+
+// NodeGroupForNode returns the node group for the given node, nil if the node
+// should not be processed by cluster autoscaler, or non-nil error if such
+// occurred. Must be implemented.
+func (d *digitaloceanCloudProvider) NodeGroupForNode(node *apiv1.Node) (cloudprovider.NodeGroup, error) {
+	nodeID, ok := node.Labels[nodeIDLabel]
+	if !ok {
+		// CA creates fake node objects to represent upcoming VMs that haven't
+		// registered as nodes yet. They have node.Spec.ProviderID set. Use
+		// that as nodeID.
+		nodeID = node.Spec.ProviderID
+	}
+
+	klog.V(5).Infof("checking nodegroup for node ID: %q", nodeID)
+
+	// NOTE(arslan): the number of node groups per cluster is usually very
+	// small. So even though this looks like quadratic runtime, it's OK to
+	// proceed with this.
+	for _, group := range d.manager.nodeGroups {
+		klog.V(5).Infof("iterating over node group %q", group.Id())
+		nodes, err := group.Nodes()
+		if err != nil {
+			return nil, err
+		}
+
+		for _, node := range nodes {
+			klog.V(5).Infof("checking node have: %q want: %q", node.Id, nodeID)
```

Comment:
This will produce incredible amount of logs in a decently sized cluster. There are some V(5) logs that are sometimes useful for debugging, but this make V(5) unusable - can you make it V(6) or more?

Reply:
Made it V(6).